core/src/main/java/org/apache/iceberg/PartitionStats.java

```diff
@@ -249,4 +250,45 @@ public <T> void set(int pos, T value) {
         throw new UnsupportedOperationException("Unknown position: " + pos);
     }
   }
+
+  @Override
+  @SuppressWarnings("checkstyle:CyclomaticComplexity")
+  public boolean equals(Object other) {
+    if (this == other) {
+      return true;
+    } else if (!(other instanceof PartitionStats)) {
+      return false;
+    }
+
+    PartitionStats that = (PartitionStats) other;
+    return Objects.equals(partition, that.partition)
```

Comment:
StructLike doesn't have equal, I think you need to use StructLikeComparator here

Reply:
We are storing `Record` as `Partition` from `coercedPartitionRecord`. 
Since `GenericRecord` has `equals` implemented, calling `Objects.equals` is working here. 
Hence, I didn't add comparator logic. 

But agree that need to understand the implementation logic, just by looking at this class, it looks like we need comparator logic. I can update it if it is necessary.

Reply:
I also tried adding the comparator logic today and pass the comparator of partition type. 

Since, we are converting the partition values for the writer in `PartitionStatsHandler.statsToRecords()`.  
Comparator expects integer value for date column but we have converted the values to LocalDate, hence comparison fails. 

If I don't use the comparator, Record.equals() will be called which does array compare and passes.

Reply:
Agree. I have added the assert (preconditions) to make sure it is always of the type Record. Also, added a comment that why keeping the type as StructType instead of Record when it is always a record. 

It is because in future when we introduce internal parquet writers that works with StructLike instead of Record, we don't have to change method signatures and it will be compatible.